dragonfly/block/door.go

```diff
@@ -0,0 +1,195 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/dragonfly/block/model"
+	"github.com/df-mc/dragonfly/dragonfly/block/wood"
+	"github.com/df-mc/dragonfly/dragonfly/item"
+	"github.com/df-mc/dragonfly/dragonfly/world"
+	"github.com/df-mc/dragonfly/dragonfly/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+)
+
+// Door is a block that can be used as an openable 1x2 barrier.
+type Door struct {
+	noNBT
+	transparent
+
+	// Wood is the type of wood of the door. This field must have one of the values found in the material
+	// package.
+	Wood wood.Wood
+	// Facing is the direction the door is facing.
+	Facing world.Direction
+	// Open is whether or not the door is open.
+	Open bool
+	// Top is whether the block is the top or bottom half of a door
+	Top bool
+	// Right is whether the door hinge is on the right side
+	Right bool
+}
+
+// Model ...
+func (d Door) Model() world.BlockModel {
+	return model.Door{Facing: d.Facing, Open: d.Open, Right: d.Right}
+}
+
+// NeighbourUpdateTick ...
+func (d Door) NeighbourUpdateTick(pos, changedNeighbour world.BlockPos, w *world.World) {
+	if d.Top {
+		if _, ok := w.Block(pos.Side(world.FaceDown)).(Door); !ok {
+			w.BreakBlock(pos)
+		}
+	} else {
+		if solid := w.Block(pos.Side(world.FaceDown)).Model().FaceSolid(pos.Side(world.FaceDown), world.FaceUp, w); !solid {
+			w.BreakBlock(pos)
+		} else if _, ok := w.Block(pos.Side(world.FaceUp)).(Door); !ok {
+			w.BreakBlock(pos)
+		}
+	}
+}
+
+// UseOnBlock handles the directional placing of doors
+func (d Door) UseOnBlock(pos world.BlockPos, face world.Face, clickPos mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	pos, face, used := firstReplaceable(w, pos, face, d)
+	if !used {
+		return false
+	}
+	if face != world.FaceUp {
+		return false
+	}
+	if solid := w.Block(pos.Side(world.FaceDown)).Model().FaceSolid(pos.Side(world.FaceDown), world.FaceUp, w); !solid {
+		return false
+	}
+	if _, ok := w.Block(pos.Side(world.FaceUp)).(Air); !ok {
+		return false
+	}
+	d.Facing = user.Facing()
+	left := w.Block(pos.Side(d.Facing.Rotate90().Opposite().Face()))
+	right := w.Block(pos.Side(d.Facing.Rotate90().Face()))
+	if door, ok := left.(Door); ok {
+		if door.Wood == d.Wood {
+			d.Right = true
+		}
+	}
+	if diffuser, ok := right.(LightDiffuser); !ok || diffuser.LightDiffusionLevel() != 0 {
+		if diffuser, ok := left.(LightDiffuser); ok && diffuser.LightDiffusionLevel() == 0 {
```

Comment:
Any reason to use LightDiffuser here? Maybe we can change this to a proper solid check.